builtin/docker/pull/kaniko.go

```diff
@@ -0,0 +1,158 @@
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	empty "google.golang.org/protobuf/types/known/emptypb"
+	"net"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+)
+
+func (b *Builder) pullWithKaniko(
+	ctx context.Context,
+	ui terminal.UI,
+	sg terminal.StepGroup,
+	log hclog.Logger,
+	ai *wpdocker.AccessInfo,
+) (*wpdocker.Image, error) {
+	step := sg.Add("Pulling Docker image with kaniko...")
+	defer func() {
+		if step != nil {
+			step.Abort()
+		}
+	}()
+
+	target := &wpdocker.Image{
+		Image:    b.config.Image,
+		Tag:      b.config.Tag,
+		Location: &wpdocker.Image_Docker{Docker: &empty.Empty{}},
+	}
+
+	var oci ociregistry.Server
+	oci.DisableEntrypoint = b.config.DisableCEB
+	oci.Logger = log
+
+	if ai.Auth != nil {
+		switch sv := ai.Auth.(type) {
+		case *wpdocker.AccessInfo_Encoded:
+			user, pass, err := wpdocker.CredentialsFromConfig(sv.Encoded)
+			if err != nil {
+				return nil, err
+			}
+			oci.AuthConfig.Username = user
+			oci.AuthConfig.Password = pass
+		case *wpdocker.AccessInfo_Header:
+			oci.AuthConfig.Auth = sv.Header
+		case *wpdocker.AccessInfo_UserPass_:
+			oci.AuthConfig.Username = sv.UserPass.Username
+			oci.AuthConfig.Password = sv.UserPass.Password
+		}
+	}
+
+	// Determine the host that we're setting auth for. We have to parse the
+	// image for this cause it may not contain a host. Luckily Docker has
+	// libs to normalize this all for us.
+	log.Trace("determining host for auth configuration", "image", target.Name())
+	ref, err := reference.ParseNormalizedNamed(target.Image)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "unable to parse image name: %s", err)
+	}
+
+	host := reference.Domain(ref)
+	if host == "docker.io" {
+		// The normalized name parse above will turn short names like "foo/bar"
+		// into "docker.io/foo/bar" but the actual registry host for these
+		// is "index.docker.io".
+		host = "index.docker.io"
+	}
+	log.Trace("auth host", "host", host)
+
+	if ai.Insecure {
+		oci.Upstream = "http://" + host
+	} else {
+		oci.Upstream = "https://" + host
+	}
+
+	refPath := reference.Path(ref)
+
+	if !b.config.DisableCEB {
+		// For Kaniko we can use our runtime arch because the image we build
+		// always matches the architecture of our Kaniko environment.
+		assetName, ok := assets.CEBArch[runtime.GOARCH]
+		if !ok {
+			return nil, status.Errorf(codes.FailedPrecondition,
+				"automatic entrypoint injection not supported for architecture: %s", runtime.GOARCH)
+		}
+
+		data, err := assets.Asset(assetName)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "unable to restore custom entry point binary: %s", err)
+		}
+
+		step.Done()
+		step = sg.Add("Testing registry and uploading entrypoint layer")
+
+		err = oci.SetupEntrypointLayer(refPath, data)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "error setting up entrypoint layer: %s", err)
+		}
+	}
+
+	li, err := net.Listen("tcp", "localhost:0")
```

Comment:
A note why we're setting up a net listener would be good.

I think the `localhost:0` means that it takes any available port, right? Nice 👍🏻